src/geopackage/sqlite/database.cpp

```diff
@@ -0,0 +1,111 @@
+#include "NGen_SQLite.hpp"
+
+using namespace geopackage;
+
+/**
+ * Get a runtime error based on a function and code.
+ *
+ * @param f String denoting the function where the error originated
+ * @param code sqlite3 result code
+ * @param extra additional messages to add to the end of the error
+ * @return std::runtime_error
+ */
+std::runtime_error sqlite_error(const std::string& f, int code, const std::string& extra = "")
+{
+    std::string errmsg = f + " returned code "
+                           + std::to_string(code)
+                           + " (msg: "
+                           + std::string(sqlite3_errstr(code))
+                           + ")";
+
+    if (!extra.empty()) {
+      errmsg += " ";
+      errmsg += extra;
+    }
+
+    return std::runtime_error(errmsg);
+}
+
+sqlite::sqlite(const std::string& path)
+{
+    sqlite3* conn;
+    int      code = sqlite3_open_v2(path.c_str(), &conn, SQLITE_OPEN_READONLY, NULL);
+    if (code != SQLITE_OK) {
+        throw sqlite_error("sqlite3_open_v2", code);
+    }
+    this->conn = sqlite_t(conn);
+}
+
+sqlite3* sqlite::connection() const noexcept
+{
+    return this->conn.get();
+}
+
+bool sqlite::has_table(const std::string& table) noexcept
+{
+    auto q = this->query("SELECT EXISTS(SELECT 1 from sqlite_master WHERE type='table' AND name=?)", table);
+    q.next();
+    return q.get<int>(0);
+};
+
+sqlite_iter sqlite::query(const std::string& statement)
+{
+    sqlite3_stmt* stmt;
+    const auto    cstmt = statement.c_str();
+    const int     code  = sqlite3_prepare_v2(this->connection(), cstmt, statement.length() + 1, &stmt, NULL);
+
+    if (code != SQLITE_OK) {
+        // something happened, can probably switch on result codes
+        // https://www.sqlite.org/rescode.html
+        throw sqlite_error("sqlite3_prepare_v2", code, "(query: " + statement + ")");
```

Comment:
Again on the "if this is part of an internet-accessible-service" front, concatenating a statement (especially one we know has failed to execute) into what may be rendered as output could result in a vulnerability.

Reply:
This is a good point, this should've been debug output (at most), or removed entirely. Will adjust :+1:

Reply:
Removed statement output in a6c57e0.

Reply:
Goes back to the need for a more production-ready logging capability... the statement should be included in one logging path and the message returned to the user in another (maybe the exception text, maybe not).

Reply:
Maybe https://github.com/gabime/spdlog could be something we look at? Not sure if there's any better logging implementations, but that's the one I'm familiar with (and it's MIT license + header only)